packages/backend/server/src/core/doc-service/service.ts

```diff
@@ -0,0 +1,97 @@
+import { FactoryProvider, Injectable, Logger } from '@nestjs/common';
+import { ModuleRef } from '@nestjs/core';
+import { ClsService } from 'nestjs-cls';
+
+import {
+  Config,
+  CryptoHelper,
+  InternalServerError,
+  UserFriendlyError,
+} from '../../base';
+import { PgWorkspaceDocStorageAdapter } from '../doc';
+import { type DocRecord } from '../doc/storage';
+
+export abstract class DocService {
+  abstract getDoc(
+    workspaceId: string,
+    docId: string
+  ): Promise<DocRecord | null>;
+}
+
+@Injectable()
+class DatabaseDocService extends DocService {
+  constructor(protected readonly workspace: PgWorkspaceDocStorageAdapter) {
+    super();
+  }
+
+  async getDoc(workspaceId: string, docId: string): Promise<DocRecord | null> {
+    return await this.workspace.getDoc(workspaceId, docId);
+  }
+}
+
+@Injectable()
+class RpcDocService extends DocService {
+  private readonly logger = new Logger(DocService.name);
+
+  constructor(
+    private readonly config: Config,
+    private readonly crypto: CryptoHelper,
+    private readonly cls: ClsService
+  ) {
+    super();
+  }
+
+  async getDoc(workspaceId: string, docId: string): Promise<DocRecord | null> {
+    const url = `${this.config.docService.endpoint}/rpc/workspaces/${workspaceId}/docs/${docId}`;
+    const accessToken = `${docId},${this.crypto.sign(docId)}`;
```

Comment:
we can update the return of `sign` to auto include the data signed to make the calls simpler
```ts
const signature = this.crypto.sign(docId) // `${content},${sig}`

this.crypto.verify(signature)

```

Reply:
Just realized crypto.sign hasn't been used, so I'll refactor it.